### PR TITLE
Update path creation and implemented possibility save data after removing PV

### PR DESCRIPTION
**Pull request by Bledai**
Hello,

This PR resolved issue #6 .

In the scope of this PR was added the ability to use metadate (name, namespace, labels, annotations) PVC field for dynamical creation the pvPath and possibility saving data on nfs-storage after deletion PV.

Example:

StorageClass:
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: nfs
provisioner: nfs.lgi.io/dynamic-provisioner
parameters:
  onDelete: "retain"
  pathPattern: "${.PVC.namespace}/${.PVC.annotations.nfs.lgi.io/storage-path}"
mountOptions:
  - lookupcache=pos
```

PersistentVolumeClaim:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: hollow-producer
  namespace: lab5a
  annotations:
    volume.beta.kubernetes.io/storage-class: "nfs"
    nfs.lgi.io/storage-path: "hollow-export" # path which will created: /lab5a/hollow-export
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 500Gi
```

### CODE REVIEW

File: cmd/nfs-subdir-external-provisioner/provisioner.go [hunk 1]
```diff
@@ -47,13 +47,36 @@ type nfsProvisioner struct {
...
 }
 
+type pvcMetadata struct {
+	data        map[string]string
+	labels      map[string]string
+	annotations map[string]string
+}
+
+func (meta *pvcMetadata) stringParser(str string) string {
+	pattern := regexp.MustCompile(`{pvc\.((labels|annotations)\.(.*?)|.*?)}`)
```
Comment (jackielii) [inline cmd/nfs-subdir-external-provisioner/provisioner.go:57]: if the pattern is static, better to move it to global scope

File: cmd/nfs-subdir-external-provisioner/provisioner.go [hunk 2]
```diff
@@ -47,13 +47,36 @@ type nfsProvisioner struct {
...
+	result := pattern.FindAllStringSubmatch(str, -1)
+	for _, r := range result {
+		switch r[2] {
+		case "labels":
+			str = strings.Replace(str, r[0], meta.labels[r[3]], -1)
+		case "annotations":
+			fmt.Println(r[0], r[3], meta.annotations[r[3]])
```
Comment (jackielii) [inline cmd/nfs-subdir-external-provisioner/provisioner.go:64]: remove

File: cmd/nfs-subdir-external-provisioner/provisioner.go [hunk 3]
```diff
@@ -112,21 +152,20 @@ func (p *nfsProvisioner) Delete(volume *v1.PersistentVolume) error {
...
-	if exists {
-		archiveBool, err := strconv.ParseBool(archiveOnDelete)
-		if err != nil {
-			return err
-		}
-		if !archiveBool {
-			return os.RemoveAll(oldPath)
-		}
-	}
+	onDelete := storageClass.Parameters["onDelete"]
```
Comment (jackielii) [inline cmd/nfs-subdir-external-provisioner/provisioner.go:155]: this would be backward incompatible

**Review by jackielii — COMMENTED** [review]
This PR tries to address multiple issue in one go and backward incompatible. I suggest you address one issue at a time. 

And also the elephant in the room: we don't have tests

**Comment by moikot** [comment]
What is the difference between Storage Class "reclaimPolicy: Retain" and introduced onDelete: "retain" ?

File: cmd/nfs-subdir-external-provisioner/provisioner.go (hunk 1 above)
Comment (Bledai) [reply cmd/nfs-subdir-external-provisioner/provisioner.go:57]: Got it

**Comment by Bledai** [comment]
> This PR tries to address multiple issue in one go and backward incompatible. I suggest you address one issue at a time.
> 
> And also the elephant in the room: we don't have tests
We opened this PR using the old address and the tests were passed.

**Comment by Bledai** [comment]
> What is the difference between Storage Class "reclaimPolicy: Retain" and introduced onDelete: "retain" ?


Now, by default, in any case, the nfs-exporter deletes the data from nfs share or move it. `"reclaimPolicy: Retain"` - this really saves the data, however, there is also a one-to-one connection between pv and pvc, which causes problems in the case of generating a new pv along the old mount path - some steps need to be done manually.
https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain
`onDelete: "retain"` In this case, it simply removes the PV entity from the kubernetes without touching the directory on the ball. 
So.. a similar name may be misleading.

File: cmd/nfs-subdir-external-provisioner/provisioner.go (hunk 2 above)
Comment (Bledai) [reply cmd/nfs-subdir-external-provisioner/provisioner.go:64]: Confused me

**Comment by Bledai** [comment]
@jackielii I can split this PR into two, if it is more convenient, and discuss the resolution of issues one by one.
What do you think about this?

File: cmd/nfs-subdir-external-provisioner/provisioner.go (hunk 3 above)
Comment (Bledai) [reply cmd/nfs-subdir-external-provisioner/provisioner.go:155]: @jackielii , not really, the default move solution is left. We wanted to propose adding a new parameter for similar operations.
However, if this is indeed a problem, the solution is to save the two parameters for full integration.

**Comment by moikot** [comment]
For some reason, I thought that the Delete method is not called at all if reclaimPolicy is "Retain"

**Comment by Bledai** [comment]
> For some reason, I thought that the Delete method is not called at all if reclaimPolicy is "Retain"

And there is. However, at the same time, the PV itself remains saved and tied to the PVС, and since the PV name is generated by the nfs-provisioner each time through the PVC's ID, this causes the presence of duplicated PVs looking at the same directory on nfs-share.

**Comment by moikot** [comment]
@Bledai Thanks a lot for your explanation! Now I see that it is quite good to have the default reclaim policy (Delete) yet having an ability to reuse the subdirs. I guess it would be nice to have such peculiarities explained somewhere in detail.

**Comment by Bledai** [comment]
> @Bledai Thanks a lot for your explanation! Now I see that it is quite good to have the default reclaim policy (Delete) yet having an ability to reuse the subdirs. I guess it would be nice to have such peculiarities explained somewhere in detail.

I think I should update the README in this regard.

**Comment by varkey** [comment]
Does this also support the use case where you'd want to mount a custom path already existing in the NFS file system?

**Comment by Bledai** [comment]
> Does this also support the use case where you'd want to mount a custom path already existing in the NFS file system?

Yes

**Comment by Bledai** [comment]
/assign @jsafrane

README has been updated and the old parameter `archiveOnDelete` has been retained, but it will be ignored if `onDelete` parameter is set.

File: README.md [hunk 4]
```diff
@@ -107,6 +107,16 @@ spec:
 
 You may also want to change the PROVISIONER_NAME above from ``fuseim.pri/ifs`` to something more descriptive like ``nfs-storage``, but if you do remember to also change the PROVISIONER_NAME in the storage class definition below:
 
+**Step 5: Deploying your storage class**
+
+***Paraments:***
```
Comment (kmova) [inline README.md:112]: n_: should be `parameters`

File: cmd/nfs-subdir-external-provisioner/provisioner.go [hunk 5]
```diff
@@ -109,24 +149,39 @@ func (p *nfsProvisioner) Delete(volume *v1.PersistentVolume) error {
...
 	// Otherwise, archive it.
 	archiveOnDelete, exists := storageClass.Parameters["archiveOnDelete"]
 	if exists {
-		archiveBool, err := strconv.ParseBool(archiveOnDelete)
-		if err != nil {
-			return err
-		}
-		if !archiveBool {
-			return os.RemoveAll(oldPath)
+		if exists {
```
Comment (kmova) [inline cmd/nfs-subdir-external-provisioner/provisioner.go:171]: why is this check repeating - 171 and 172?

**Comment by kmova** [comment]
@Bledai - can you also add a new document on how this feature can be used to create a new PVC/PV with archived data?

File: cmd/nfs-subdir-external-provisioner/provisioner.go (hunk 5 above)
Comment (Bledai) [reply cmd/nfs-subdir-external-provisioner/provisioner.go]: removed

File: README.md (hunk 4 above)
Comment (Bledai) [reply README.md]: done

**Comment by Bledai** [comment]
> @Bledai - can you also add a new document on how this feature can be used to create a new PVC/PV with archived data?

@kmova. You mean to add an example description for CtorageClass for archived date cases or something different? Since it has already been described that this is the default behavior, we can also add that the archived data is always will be stored in the root path for nfs-share. - I left this part unchanged

File: README.md [hunk 6]
```diff
@@ -109,7 +109,7 @@ You may also want to change the PROVISIONER_NAME above from ``fuseim.pri/ifs`` t
 
 **Step 5: Deploying your storage class**
 
-***Paraments:***
+***paraments:***
```
Comment (kmova) [inline README.md:112]:
```suggestion
***Parameters:***
```

**Comment by kmova** [comment]
> @kmova. You mean to add an example description for CtorageClass for archived date cases or something different?

Yes, it is about the archived data case. This PR describes how to archive the data. I was wondering if you also can provide some documentation on how to use the archived data to create a new PV. Note that this doesn't need to be in the same README. Mostly trying to get more details on how you want to use this feature in your clusters.

**Comment by Bledai** [comment]
> Yes, it is about the archived data case. This PR describes how to archive the data. I was wondering if you also can provide some documentation on how to use the archived data to create a new PV. Note that this doesn't need to be in the same README. Mostly trying to get more details on how you want to use this feature in your clusters.

@kmova A bit confused by your request since the main purpose of this PR is to impregnate the ability to create a custom path and reuse previously created directories for PVC/PV. 
We did not think about archiving data in our clusters - since the archive is created with an unreadable name - since the PV itself is created with that:
https://github.com/kubernetes-csi/external-provisioner/blob/abf62116073f85f6bb5be34a4a80666c29a28dee/pkg/controller/controller.go#L500
https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/pull/7/files#diff-ea36a5773b4590705db28106e082a8bbR180 (we kept the logic of creating the archive intact)
It is quite difficult to plan something when the archive is created with a random name (`options.PVC.ObjectMeta.UID`).
Of course, this is done for the sake of the uniqueness of the name. 
It seems to me that it is worth thinking about the possibility of setting the archive name through the parameters of the StorageClass.
However, this PR is not about that.

**Comment by kmova** [comment]
Thanks for the explanation @Bledai. @jackielii - do you have any further comments on this PR?

**Comment by jackielii** [comment]
LGTM

**Comment by Bledai** [comment]
@kmova @jackielii waiting for approval.

**Comment by Bledai** [comment]
@jsafrane please take a look

**Comment by kmova** [comment]
Like @jackielii mentioned, this repo badly needs to have tests, which can be taken up in the upcoming PR and also will have to take up covering a few cases/examples where:
- StorageClass is specified without the new parameters
- StorageClass is specified with only archiveOnDelete parameter set to true.

**Comment by sunliusi** [comment]
Have you implemented "mount custom paths"?

It doesn't work after my test

**Pull request merged**